### PR TITLE
Updated PyChiquito frontend for latest changes

### PR DESCRIPTION
**Pull request by qwang98**
Most specifically, need to acquire num_step from ast to calculate padding in PyChiquito.

### CODE REVIEW

**Comment by leolara** [comment]
@qwang98 I don't understand, the num_steps is set in the python frontend, and the padding is done in the python frontend. So I don't understand what you mean about getting num_steps from AST. The AST is in the python frontend where the padding is done.

**Comment by qwang98** [comment]
Sorry for the confusion. My description was incorrect and doesn't match the changes. The changes are:
1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.
3. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
4. Added empty trace function for AST Circuit deserializer, because otherwise the compiler won't generate witness if trace defaults to None.

@leolara Ready for review again :)

**Comment by leolara** [comment]
> 1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.

So the solution for this is explained here: https://github.com/privacy-scaling-explorations/chiquito/issues/83 and also the exposed column is in the plonkish IR and in the ChiquitoHalo2 so it wouldn't be necessary to keep the AST for that.

**Comment by leolara** [comment]
> 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.

Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend

**Comment by leolara** [comment]
> 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.

I think this is better solved by implementing Debug and/or Display traits on certain types.

**Comment by qwang98** [comment]
> > 1. Store AST in the CIRCUIT_MAP, so that the MockProver can obtain whether there's exposed columns in the AST, and fill in the third `instance` parameter of `MockProver::run` accordingly, i.e. empty Vec if there's no instance or non-empty Vec if there is.
> 
> So the solution for this is explained here: #83 and also the exposed column is in the plonkish IR and in the ChiquitoHalo2 so it wouldn't be necessary to keep the AST for that.

Cool. Working on this now :)

**Comment by qwang98** [comment]
> > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> 
> Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend

It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.

**Comment by qwang98** [comment]
> > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> 
> I think this is better solved by implementing Debug and/or Display traits on certain types.

I see your point now. I'll think about it and either implement this or remove the print_plaf_witness function entirely. :)

**Comment by leolara** [comment]
The optimal solution would be to implement unit tests

**Comment by leolara** [comment]
> > > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> > 
> > 
> > Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend
> 
> It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.

We should implement the Debug and Display for TraceWitness struct, Assigments, etc... much easier that way. But can be done in another PR

**Comment by leolara** [comment]
@qwang98 I think we can make it for the super circuit, if we follow the same order as the meta.instance_columns are called.

File: src/plonkish/backend/halo2.rs [hunk 1]
```diff
@@ -388,13 +415,35 @@ pub struct ChiquitoHalo2SuperCircuit<F: Field + From<u64>> {
...
     pub fn new(sub_circuits: Vec<ChiquitoHalo2<F>>, witness: SuperAssignments<F>) -> Self {
         Self {
             sub_circuits,
             witness,
         }
     }
+
+    pub fn instance(&self) -> Vec<Vec<F>> {
+        if !self.witness.is_empty() {
+            let mut result = Vec::new();
```
Comment (leolara) [inline src/plonkish/backend/halo2.rs:428]: if you put let mut result outside if, you don't need the else. Just write result in the last line

File: src/plonkish/backend/halo2.rs [hunk 2]
```diff
@@ -388,13 +415,35 @@ pub struct ChiquitoHalo2SuperCircuit<F: Field + From<u64>> {
...
+
+            for (uuid, assignment) in &self.witness {
```
Comment (leolara) [inline src/plonkish/backend/halo2.rs:430]: @qwang98 I don't think you should follow the order of the witness, but the order of sub_circuit, or whatever order is used to call meta.instance_column()

Comment (qwang98) [reply src/plonkish/backend/halo2.rs:430]: Agreed. sub_circuit is the order meta.instance_column is called.

File: src/plonkish/backend/halo2.rs (hunk 1 above)
Comment (qwang98) [reply src/plonkish/backend/halo2.rs:428]: Indeed.

**Comment by qwang98** [comment]
> @qwang98 I think we can make it for the super circuit, if we follow the same order as the meta.instance_columns are called.

Done

**Comment by qwang98** [comment]
> > > > 2. Added a debug function that prints witness after converting it to Plaf. This was quite helpful for identifying a couple bugs that I fixed. I exposed it in PyO3 but not in Python DSL, as it's not some function the user will use.
> > > 
> > > 
> > > Why is there plaf stufff in the python frontend? I think for now we should not implement plaf backend for the python frontend
> > 
> > 
> > It's not implementing Plaf backend. It's really just printing witness assignment with the help of Plaf.
> 
> We should implement the Debug and Display for TraceWitness struct, Assigments, etc... much easier that way. But can be done in another PR

Agree on this.

**Review by leolara — APPROVED** [review]

**Pull request merged**